host/volume/zfs/zfs.go

```diff
@@ -0,0 +1,181 @@
+package zfs
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"time"
+
+	zfs "github.com/flynn/flynn/Godeps/_workspace/src/github.com/mistifyio/go-zfs"
+	"github.com/flynn/flynn/host/volume"
+	"github.com/flynn/flynn/pkg/random"
+)
+
+type zfsVolume struct {
+	info   *volume.Info
+	mounts map[volume.VolumeMount]struct{}
+	// FIXME: starting to look better to put this back in the hands of the provider, and making all of the challenges of maintaining entanglement with external state confined to that.
+	poolName  string // The name of the zpool this storage is cut from.  (We need this when forking snapshots, or doing some inspections.)
+	basemount string // This is the location of the main mount of the ZFS dataset.  Mounts into containers are bind-mounts pointing back out to this.  The user does not control it (it is essentially an implementation detail).
+}
+
+type Provider struct {
+	config  *ProviderConfig
+	dataset *zfs.Dataset
+}
+
+/*
+	Describes zfs config used at provider setup time.
+
+	`volume.ProviderSpec.Config` is deserialized to this for zfs.
+*/
+type ProviderConfig struct {
+	// DatasetName specifies the zfs dataset this provider will create volumes under.
+	//
+	// If it doesn't specify an existing dataset, and `MakeDev` parameters have
+	// been provided, those will be followed to create a zpool; otherwise
+	// provider creation will fail.
+	DatasetName string `json:"dataset"`
+
+	Make *MakeDev `json:"makedev,omitempty"`
+}
+
+/*
+	Describes parameters for creating a zpool.
+
+	Currently this only supports file-type vdevs; be aware that these are
+	convenient, but may have limited performance.  Advanced users should
+	consider configuring a zpool using block devices directly, and specifing
+	use of datasets in those zpools those rather than this fallback mechanism.
+*/
+type MakeDev struct {
+	BackingFilename string `json:"filename"`
+	Size            int64  `json:"size"`
+}
+
+func NewProvider(config *ProviderConfig) (volume.Provider, error) {
+	if _, err := exec.LookPath("zfs"); err != nil {
+		return nil, fmt.Errorf("zfs command is not available")
+	}
+	dataset, err := zfs.GetDataset(config.DatasetName)
+	if err != nil {
+		err = eunwrap(err)
+		switch err.Error() {
+		case fmt.Sprintf("cannot open '%s': dataset does not exist\n", config.DatasetName):
```

Comment:
Can we just match against `dataset does not exist` instead of the exact string? This seems like it could break across versions of the zfs userspace tools pretty easily.